File: mergesort.py
a = [9, 2, 4, 2, 5, 3, 1, 8, 7, 6]

#merge module
def merge (b, c) :
	bl = len(b)
	cl = len(c)
	if bl == 0 :
		return c
	if cl == 0 :
		return b
	d = [] 
	i = 0
	j = 0
	k = 0
#elements copied to new list	
	while j < bl and k < cl :
		if b[j] < c[k] :
			d.append(b[j])
			j = j + 1
		else :
			d.append(c[k])
			k = k + 1
#check for remaining elements in lists
	if j < bl :
			d += b[j:]
		
	if k < cl :
			d += c[k:]
		 		
	return d		

#merge sort	
def merge_sort (arr) :
	l  = len(arr)
	d = []
	if l > 1 :
		arr1 = arr[:l/2]
		arr2 = arr[l/2 : ]
		b =  merge_sort (arr1) 
		c =  merge_sort (arr2) 
		d =  merge (b, c) 
		return d
	else :	
		return arr
	
print merge_sort (a)
print a

File: mergeSort.c
#include <stdio.h>


void merge(int a[],int p, int q, int r){
	int c[10],b[10];
	int i,j,k,l;
	j =0;k=0;
	int n1,n2;
  	n1=q-p+1;
 	n2=r-q;

  for(i=0; i<n1; i++)
    b[i]=a[p+i];
  for(j=0; j<n2; j++)
    c[j]=a[q+j+1];

	b[i] = 9999;
	c[j] = 9999;
	i=j=0;
	for(k=p;k<=r;k++) {
		if (c[i]<=b[j])
		{
			a[k]=c[i];
			i++;
		}
		else {
			a[k] = b[j];
			j++;
		}
	}



}

void mergesort(int a[],int p, int r) {

	if (p < r) {
		int q = (p+r)/2;
		mergesort(a,p,q);
		mergesort(a,q+1,r);
		merge(a,p,q,r);
	}
	
} 

int main(int argc, char const *argv[])
{
	int a[] = {7,6,5,4,3,2,1,0};
	int p =0, r = 7,i;	
	mergesort(a,p,r);

	for (i = 0; i < 8; ++i) {

		printf("%d \n",a[i] );
	}
	return 0;
}